pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -683,6 +683,8 @@ func (gb *GraphBuilder) processGraphChanges() bool {
 			gb.addUnblockedOwnersToDeleteQueue(removed, changed)
 			// update the node itself
 			existingNode.owners = accessor.GetOwnerReferences()
+			// Make sure the API version stored in memory is up to date.
+			existingNode.identity.APIVersion = event.gvk.GroupVersion().String()
```

Comment:
A few observations:

1. This code is only run in the case where the add/update modifies ownerReferences of the observed object (it runs in the `if len(added) != 0 || len(removed) != 0 || len(changed) != 0 {` block)... it won't run if the only thing changing is the apiVersion/kind of the observed object

2. `identity` isn't safe to update, since other threads read from this field and assume it is immutable... we would have to do something like what is done above when transitioning from an unobserved node to an observed node:

```
existingNode = existingNode.clone()
existingNode.identity.APIVersion = event.gvk.GroupVersion().String()
existingNode.identity.Kind = event.gvk.Kind
gb.uidToNode.Write(existingNode)
```

3. We should only do this swap when the observed apiVersion/kind doesn't match the existing identity, and the event we're observing is not a virtual event

4. This deserves testcases added to TestConflictingData